C++: Output iterators in AST taint tracking

rdmarsh2 opened the pull request:
This is a first cut at output iterators in AST taint tracking, implemented by adding new edges from (direct) assignments to `*iter` to any reads of the collection it's derived from. This will have a fairly high rate of false positives, and the next step would be modifying `FlowVar.qll` to add `PostUpdateNode`s that could allow more precision.

Note that this doesn't handle range-based for loops, because the reference from `*iter` is stored in the loop variable rather than being assigned into immediately.

MathiasVP commented on cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:
```diff
@@ -0,0 +1,342 @@
+  IteratorAddOperator() {
+    this.hasName("operator+") and
+    (
+      this
+          .getACallToThisFunction()
+          .getArgument(0)
+          .getFullyConverted()
+          .getUnspecifiedType()
+          .(PointerType)
+          .getBaseType() instanceof Iterator or
+      this
+          .getACallToThisFunction()
+          .getArgument(0)
+          .getFullyConverted()
+          .getUnspecifiedType()
+          .(PointerType)
+          .getBaseType() instanceof Iterator
```
Are one of those `0`s supposed to be a `1`? They look really identical to me! If that's the correct fix, what about writing this as:

```codeql
this
    .getACallToThisFunction()
    .getArgument([0, 1])
    .getFullyConverted()
    .getUnspecifiedType()
    .(PointerType)
    .getBaseType() instanceof Iterator
```

MathiasVP commented on cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:
```diff
@@ -0,0 +1,342 @@
+      this.hasName("operator+=") or
+      this.hasName("operator-=")
+    ) and
+    this.getDeclaringType() instanceof Iterator
```
If this a non-member (as the qldoc says) then I don't think this conjunct will hold. Will it?

jbj commented:
@MathiasVP Most of the commits in this PR come from #4102. I think the first new commit is "C++: simple tests for vector output iterators".

jbj commented on cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:
```diff
@@ -324,3 +324,15 @@ class IteratorArrayMemberOperator extends MemberFunction, TaintFunction {
+class BeginOrEndFunction extends MemberFunction, TaintFunction {
+  BeginOrEndFunction() {
+    this.hasName(["begin", "cbegin", "rbegin", "crbegin", "end", "cend", "rend", "crend"]) and
+    this.getType().getUnspecifiedType() instanceof Iterator
```
```suggestion
    this.getUnspecifiedType() instanceof Iterator
```

This is shorter.

jbj commented on cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:
```diff
@@ -324,3 +324,15 @@ class IteratorArrayMemberOperator extends MemberFunction, TaintFunction {
+
+class BeginOrEndFunction extends MemberFunction, TaintFunction {
+  BeginOrEndFunction() {
+    this.hasName(["begin", "cbegin", "rbegin", "crbegin", "end", "cend", "rend", "crend"]) and
```
This is great. Have we also modelled [`std::begin` and `std::cbegin`](https://en.cppreference.com/w/cpp/iterator/begin)?

geoffw0 replied:
This model can replace `StdStringBeginEnd` and most of `StdSequenceContainerBeginEnd` (in https://github.com/github/codeql/pull/4201/files#diff-cbbb968ec9678cb6b97fb7001e0f7261R173).

geoffw0 replied:
Ah, I see you've removed `StdStringBeginEnd` later in the PR.  I will take responsibility for removing `StdSequenceContainerBeginEnd` as well once both PRs have been merged.

geoffw0 replied:
`StdSequenceContainerBeginEnd ` removed in https://github.com/github/codeql/pull/4395.

jbj commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/TaintTrackingUtil.qll:
```diff
@@ -173,6 +174,12 @@ private predicate exprToExprStep(Expr exprIn, Expr exprOut) {
+  exists(Variable iterator, Variable collection |
+    assignmentViaIterator(iterator, exprIn) and
+    isIteratorForCollection(iterator, collection) and
+    collection.getAnAccess() = exprOut
```
This is certainly a step in the right direction but carries a risk of false positives and quadratic performance. Some tests showing false positives would be helpful when evaluating improvements to this code.

geoffw0 commented on cpp/ql/test/library-tests/dataflow/taint-tests/vector.cpp:
```diff
@@ -141,3 +141,30 @@ void test_vector_clear() {
+	for(std::vector<int>::iterator it = v4.begin(); it != v4.end(); ++it) {
+		*it = source();
+	}
+	sink(v4); // tainted [NOT DETECTED by IR]
```
I can't see any difference between the test with `v4 and the test with `v2`.

Apart from that these are great tests and really show the value of this PR.

geoffw0 replied:
I see that you've fixed this now - thanks!

geoffw0 reviewed (commented):
Looks valuable.  I agree with the comment @jbj made about more testing for false positives - ideally motivated by real world patterns.

geoffw0 commented on cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:
```diff
@@ -0,0 +1,342 @@
+class IteratorCrementOperator extends Operator, DataFlowFunction {
+  IteratorCrementOperator() {
+    (
+      this.hasName("operator++") or
+      this.hasName("operator--")
```
I believe they are both called `operator++` (or `operator--`) but have slightly different parameters to distinguish them (a dummy `int` parameter is added to one of them).  It's an odd little language quirk.  We test for the taint flow in these two test cases:
```
sink(*(++i3)); // tainted
```
and
```
sink(*(--i4)); // tainted
```
So I wouldn't mind seeing another two test cases added.

geoffw0 replied:
The above related to code added by an older PR, in the string.cpp taint test; assuming @rdmarsh2 hasn't addressed this I don't mind making a separate PR to add the other two cases.

It would be good to have answers to the other questions that relate to the older PR - even though it's probably not appropriate to fix things here I think they are still good questions.

geoffw0 replied:
I've added the two test missing test cases I was talking about in https://github.com/github/codeql/pull/4394.

rdmarsh2 commented:
I've rebased, switched to using `FlowVar` to reduce false positives, and added a few more tests. I think I've preserved the join order fixes from #4146

rdmarsh2 commented:
I've started a [CPP-Differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1408/) job.

geoffw0 commented on cpp/ql/test/library-tests/dataflow/taint-tests/vector.cpp:
```diff
@@ -325,3 +325,58 @@ void test_constructors_more() {
 	sink(v4); // tainted
 }
+
+void taint_vector_output_iterator(std::vector<int>::iterator iter) {
```
This doesn't seem to be used anywhere.

rdmarsh2 replied:
I've used it in the `v4` test. There isn't flow out of the call, though.

geoffw0 replied:
Thanks for fixing.

jbj commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll:
```diff
@@ -166,6 +157,95 @@ private module PartialDefinitions {
+      exists(Expr convertedInner |
+        valueToUpdate(convertedInner, this.getFullyConverted(), node) and
+        innerDefinedExpr = convertedInner.getUnconverted() and
+        not this instanceof Conversion
```
We already get `not this instanceof Conversion` from our abstract parent class, right?

jbj commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll:
```diff
@@ -147,10 +141,7 @@ private module PartialDefinitions {
-      inner = innerDefinedExpr and
-      outer = this
-    }
+    predicate definesExpressions(Expr inner, Expr outer) { none() }
```
Can these four predicates be `abstract`? Then it should be more clear that they are always overridden.

jbj commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll:
```diff
@@ -297,6 +297,18 @@ private class PartialDefinitionNode extends PostUpdateNode, TPartialDefinitionNo
   override string toString() { result = getPreUpdateNode().toString() + " [post update]" }
 }
 
+private class VariablePartialDefinitionNode extends PartialDefinitionNode {
+  override VariablePartialDefinition pd;
+
+  override Node getPreUpdateNode() { pd.definesExpressions(_, result.asExpr()) }
+}
+
+private class IteratorPartialDefinitionNode extends PartialDefinitionNode {
+  override IteratorPartialDefinition pd;
+
+  override Node getPreUpdateNode() { pd.definesExpressions(_, result.asExpr()) }
```
What are these classes used for? I expected that they'd have to do with a distinction between data flow and taint. In particular, I'd expect the use of `PartialDefinitionNode` in `simpleLocalFlowStep` to become `VariablePartialDefinitionNode` now.

jbj commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/TaintTrackingUtil.qll:
```diff
@@ -248,4 +249,18 @@ private predicate exprToPartialDefinitionStep(Expr exprIn, Expr exprOut) {
+}
+
+private predicate iteratorDereference(Call c) {
+    c.getTarget() instanceof IteratorArrayMemberOperator
+    or
+    c.getTarget() instanceof IteratorPointerDereferenceMemberOperator
+    or
+    c.getTarget() instanceof IteratorPointerDereferenceOperator
```
We don't usually include an `implementations` file directly because it's very hard to keep code like this in sync with `Iterator.qll`. Can the property of being an iterator dereference become a model interface instead?

jbj commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll:
```diff
@@ -166,6 +157,95 @@ private module PartialDefinitions {
+
   /**
    * A partial definition that's a definition by reference.
    */
```
I would have expected `DefinitionByReference`, on the line below, to now extend `VariablePartialDefinition` in order to preserve its previous behaviour.

jbj commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/TaintTrackingUtil.qll:
```diff
@@ -30,6 +31,11 @@ predicate localTaintStep(DataFlow::Node src, DataFlow::Node sink) {
 predicate defaultAdditionalTaintStep(DataFlow::Node src, DataFlow::Node sink) {
   localAdditionalTaintStep(src, sink)
+  or
+  exists(FunctionCall call, int i |
+    sink.(DataFlow::IteratorPartialDefinitionNode).getPartialDefinition().definesExpressions(_, call.getArgument(i)) and
+    src.(DataFlow::RefParameterFinalValueNode).getParameter() = call.getTarget().getParameter(i)
+  )
```
We don't usually want to add interprocedural steps here. It means we lose the calling context and that we're not using the virtual-dispatch library.

I'm afraid this is only a quick drive-by review; remind me to follow up.

rdmarsh2 replied:
Oh - in that case these steps may need to be in `DataFlow` rather than `TaintTracking`, which means reverting fa0e27b and separating `DefinitionByReferenceNode` and `RefOutNode`

jbj reviewed (commented):
Otherwise this is looking good. The remaining issue I have is with the names from `Iterator.qll` being exposed.

jbj commented on cpp/ql/src/semmle/code/cpp/models/interfaces/Iterator.qll:
```diff
@@ -0,0 +1,19 @@
+ * can be used to write to the iterator's underlying collection.
+ */
+abstract class IteratorReferenceFunction extends Function {
+  abstract FunctionInput getIteratorInput();
```
This predicate is never called.

jbj commented:
Robert has started https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1457 to evaluate this PR.

jbj commented:
https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1457 shows a 39s slowdown in `NewArrayDeleteMismatch.ql` for Wireshark. That query is the first in the suite to trigger the `FlowVar` stage, I think.

@rdmarsh2, can you find out whether that slowdown is avoidable?

rdmarsh2 commented:
There were multiple rescans of the charpred for `PartialDefinition`, which was most of the expression table. I've fixed the rescans, which should fix most of that slowdown.
